src/client.rs

```diff
@@ -141,107 +185,123 @@ impl ReconnectingRequest {
-                            )
-                        }
-                        Err(e) => return Err(Error::HttpRequest(Box::new(e))),
+                    if !resp.status().is_success() {
+                        let e = StatusError {
+                            status: resp.status(),
+                        };
+                        return Poll::Ready(Some(Err(Error::HttpRequest(Box::new(e)))));
                     }
+
+                    self.as_mut().reset_backoff();
+                    State::Connected(resp.into_body())
                 }
-                State::Connected(ref mut chunks) => match chunks.poll() {
-                    Ok(result) => {
-                        return Ok(result);
+                Err(e) => {
+                    warn!("request returned an error: {}", e);
+                    if !*retry {
+                        return Poll::Ready(Some(Err(Error::HttpStream(Box::new(e)))));
                     }
-                    Err(e) => {
-                        warn!("chunk stream returned an error: {:?}", e);
-                        if self.props.reconnect_opts.reconnect {
-                            State::WaitingToReconnect(delay(self.backoff(), "reconnecting"))
-                        } else {
-                            return Err(e);
-                        }
-                    }
-                },
-                State::WaitingToReconnect(ref mut delay) => {
-                    try_ready!(delay.poll());
-                    info!("Reconnecting");
-                    let resp = self.send_request();
-                    State::Connecting {
-                        retry: true,
-                        resp: Box::new(resp),
+                    State::WaitingToReconnect(delay(self.as_mut().backoff(), "retrying"))
+                }
+            },
+            StateProj::Connected(body) => match ready!(body.poll_data(cx)) {
+                Some(Ok(result)) => {
+                    return Poll::Ready(Some(Ok(result)));
+                }
+                res => {
+                    // reconnect
+                    if self.props.reconnect_opts.reconnect {
+                        State::WaitingToReconnect(delay(self.as_mut().backoff(), "reconnecting"))
+                    } else {
+                        return Poll::Ready(
+                            res.map(|r| r.map_err(|e| Error::HttpStream(Box::new(e)))),
+                        );
                     }
                 }
-            };
-            self.state = new_state;
-        }
+            },
+            StateProj::WaitingToReconnect(delay) => {
+                ready!(delay.poll(cx));
+                info!("Reconnecting");
+                let resp = self.send_request()?;
+                State::Connecting { retry: true, resp }
+            }
+        };
+        self.as_mut().project().state.set(new_state);
+        self.poll_next(cx)
```

Comment:
This is replacing a `loop {}` with a recursive call, correct? Is that the idiomatic way to write stream combinators in modern async Rust?

This is intuitively a little surprising to me, and according to a (very) brief Google search, [tail call optimisation is possible but not guaranteed](https://stackoverflow.com/a/59258170). We would obviously not want a possibility of stack overflow, given that these streams could be very long-lived (I believe we regularly see stream connections open for days).

Very willing to be convinced that I'm behind the times and this is a normal and reasonable thing to do :)

Similar question about `Decoded::poll_next`.

Reply:
I did some more research and found [this article](https://dev.to/seanchen1991/the-story-of-tail-call-optimizations-in-rust-35hf) which also seems to suggest that TCO is not guaranteed, so I took the liberty of changing this back to a `loop`: eebd97b